File: tests/benchmarks/python/bench.py
#!/usr/bin/env python3
import argparse, json, os, platform, socket, time
from datetime import datetime, timezone
from pathlib import Path

SCENARIOS = [
    ("standard", "standard", 0x71, 32, False),
    ("bulk", "bulk", 0x74, 512, False),
    ("network", "network", 0x78, 96, True),
]

def fletcher(data):
    a = b = 0
    for x in data:
        a = (a + x) & 0xff; b = (b + a) & 0xff
    return a, b

def encode(profile, type_byte, payload_len, network=False, seq=1):
    payload = bytes((i * 31 + payload_len) & 0xff for i in range(payload_len))
    body = bytearray()
    if network:
        body += bytes([seq & 0xff, 1, 42, payload_len & 0xff, (payload_len >> 8) & 0xff, 7, 3])
    elif profile == "bulk":
        body += bytes([payload_len & 0xff, (payload_len >> 8) & 0xff, 7, 3])
    else:
        body += bytes([payload_len & 0xff, 3])
    body += payload
    c1, c2 = fletcher(body)
    return bytes([0x90, type_byte]) + bytes(body) + bytes([c1, c2])

def decode(frame):
    if frame[0] != 0x90 or not (0x70 <= frame[1] <= 0x78):
        raise ValueError("bad sync")
    off = 2; ptype = frame[1] - 0x70
    if ptype == 8:
        off += 3; length = frame[off] | (frame[off+1] << 8); off += 3; off += 1
    elif ptype == 4:
        length = frame[off] | (frame[off+1] << 8); off += 4
    else:
        length = frame[off]; off += 2
    end = off + length
    c1, c2 = fletcher(frame[2:end])
    if frame[end] != c1 or frame[end+1] != c2:
        raise ValueError("bad crc")
    return frame[off:end]

def pct(values, q):
    values.sort(); return values[min(len(values)-1, int((len(values)-1) * q))]

def run_case(name, profile, op, iterations):
    scenario = next(s for s in SCENARIOS if s[0] == profile)
    _, *encode_args = scenario
    frames = [encode(*encode_args, seq=i) for i in range(8)]
    lat = []; total_bytes = 0; start = time.perf_counter_ns()
    for i in range(iterations):
        t0 = time.perf_counter_ns()
        if op == "encode":
            frame = encode(*encode_args, seq=i); total_bytes += len(frame)
        elif op == "decode":
            frame = frames[i % len(frames)]
            decode(frame)
            total_bytes += len(frame)
        else:
            frame = encode(*encode_args, seq=i); decode(frame); total_bytes += len(frame)
        lat.append(time.perf_counter_ns() - t0)
    duration = max((time.perf_counter_ns() - start) / 1e9, 1e-9)
    return {"name": name, "profile": profile, "operation": op, "msg_count": iterations, "bytes_total": total_bytes,
            "duration_s": duration, "msg_per_sec": iterations / duration, "mb_per_sec": total_bytes / duration / 1_000_000,
            "latency_ns": {"p50": pct(lat, .50), "p95": pct(lat, .95), "p99": pct(lat, .99), "max": max(lat)}}

def main():
    ap = argparse.ArgumentParser(); ap.add_argument("--iterations", type=int, default=int(os.getenv("BENCH_ITERATIONS", "50000"))); ap.add_argument("--output", default="tests/benchmarks/results/python.json")
    args = ap.parse_args(); scenarios=[]
    for profile, *_ in SCENARIOS:
        for op in ("encode", "decode"):
            scenarios.append(run_case(f"{profile}_{op}", profile, op, args.iterations))
    result = {"schema_version":"1", "language":"python", "runner_version":"tier-d-1", "timestamp":datetime.now(timezone.utc).isoformat().replace('+00:00','Z'),
              "host":{"os":platform.system(), "arch":platform.machine(), "cpu":platform.processor() or socket.gethostname()}, "scenarios":scenarios}
    Path(args.output).parent.mkdir(parents=True, exist_ok=True); Path(args.output).write_text(json.dumps(result, indent=2)+"\n")
if __name__ == "__main__": main()

File: tests/benchmarks/ts/bench.ts
import { writeFileSync, mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import { arch, cpus, platform } from 'node:os';
import { performance } from 'node:perf_hooks';

type Scenario = [string, string, number, number, boolean];

const scenarios: Scenario[] = [
  ['standard', 'standard', 0x71, 32, false],
  ['bulk',     'bulk',     0x74, 512, false],
  ['network',  'network',  0x78, 96,  true],
];

function fletcher(buf: Buffer): [number, number] {
  let a = 0, b = 0;
  for (const x of buf) {
    a = (a + x) & 0xff;
    b = (b + a) & 0xff;
  }
  return [a, b];
}

function verifyPayload(payload: Buffer): void {
  const payloadLen = payload.length;
  if (payloadLen === 0) return;
  const checks = new Set([0, Math.floor(payloadLen / 2), payloadLen - 1]);
  for (const i of checks) {
    const expected = (i * 31 + payloadLen) & 0xff;
    if (payload[i] !== expected) throw new Error('bad payload');
  }
}

function encode(profile: string, typeByte: number, payloadLen: number, network = false, seq = 1): Buffer {
  const payload = Buffer.alloc(payloadLen);
  for (let i = 0; i < payloadLen; i++) {
    payload[i] = (i * 31 + payloadLen) & 0xff;
  }
  const body: number[] = [];
  if (network) {
    body.push(seq & 0xff, 1, 42, payloadLen & 0xff, (payloadLen >> 8) & 0xff, 7, 3);
  } else if (profile === 'bulk') {
    body.push(payloadLen & 0xff, (payloadLen >> 8) & 0xff, 7, 3);
  } else {
    body.push(payloadLen & 0xff, 3);
  }
  const b = Buffer.concat([Buffer.from(body), payload]);
  const [c1, c2] = fletcher(b);
  return Buffer.concat([Buffer.from([0x90, typeByte]), b, Buffer.from([c1, c2])]);
}

function decode(frame: Buffer): Buffer {
  if (frame[0] !== 0x90) throw new Error('bad sync');
  let off = 2;
  const p = frame[1] - 0x70;
  let len = 0;
  if (p === 8) {
    off += 3;
    len = frame[off] | (frame[off + 1] << 8);
    off += 4;
  } else if (p === 4) {
    len = frame[off] | (frame[off + 1] << 8);
    off += 4;
  } else {
    len = frame[off];
    off += 2;
  }
  const end = off + len;
  const [c1, c2] = fletcher(frame.subarray(2, end));
  if (frame[end] !== c1 || frame[end + 1] !== c2) throw new Error('bad crc');
  const payload = frame.subarray(off, end);
  verifyPayload(payload);
  return payload;
}

function pct(v: number[], q: number): number {
  v.sort((a, b) => a - b);
  return v[Math.min(v.length - 1, Math.floor((v.length - 1) * q))];
}

function encScenario(s: Scenario, seq: number): Buffer {
  return encode(s[1], s[2], s[3], s[4], seq);
}

function runCase(name: string, profile: string, op: string, iterations: number) {
  const scenario = scenarios.find(s => s[0] === profile)!;
  const frames = Array.from({ length: 8 }, (_, i) => encScenario(scenario, i));
  const lat: number[] = [];
  let bytes = 0;
  const start = performance.now();
  for (let i = 0; i < iterations; i++) {
    const t0 = performance.now();
    if (op === 'encode') {
      const f = encScenario(scenario, i);
      bytes += f.length;
    } else {
      const f = frames[i % frames.length];
      decode(f);
      bytes += f.length;
    }
    lat.push(Math.round((performance.now() - t0) * 1e6));
  }
  const dur = Math.max((performance.now() - start) / 1000, 1e-9);
  return {
    name, profile, operation: op,
    msg_count: iterations, bytes_total: bytes, duration_s: dur,
    msg_per_sec: iterations / dur,
    mb_per_sec: bytes / dur / 1e6,
    latency_ns: {
      p50: pct(lat, 0.5), p95: pct(lat, 0.95),
      p99: pct(lat, 0.99), max: Math.max(...lat),
    },
  };
}

const args = process.argv.slice(2);
const iterations = Number(
  args[args.indexOf('--iterations') + 1] || process.env.BENCH_ITERATIONS || 50000,
);
const output = args.includes('--output')
  ? args[args.indexOf('--output') + 1]
  : 'tests/benchmarks/results/ts.json';

const out: any = {
  schema_version: '1',
  language: 'ts',
  runner_version: 'tier-d-1',
  timestamp: new Date().toISOString(),
  host: { os: platform(), arch: arch(), cpu: cpus()[0]?.model || 'unknown' },
  scenarios: [],
};
for (const [profile] of scenarios) {
  for (const op of ['encode', 'decode']) {
    out.scenarios.push(runCase(`${profile}_${op}`, profile, op, iterations));
  }
}
mkdirSync(dirname(output), { recursive: true });
writeFileSync(output, JSON.stringify(out, null, 2) + '\n');
